### ISSUE TITLE
What's the relationship between visibility and intersection?

### ISSUE THREAD

**Issue by ChumpChief** [szager-chromium/IntersectionObserver#2]
The explainer doc cites the motivation of V2 to be combating click-jacking and UI redress, in the scenario that a malicious site author tricks the user into clicking on a third-party iframe which then unwittingly accepts the click as being in good faith.

This sounds like a great problem to solve.  However, I don't think I understand its relation to IntersectionObserver.  It doesn't seem inherently related to the question of "Does this element intersect the viewport?" - rather I think the author's question is "I just received some interaction - can I trust it was received in good-faith?"

For this question, I think IO is not well-suited since it doesn't have any inherent ties to user interaction.  This IO approach requires the iframed content author to monitor events and maintain state in order to correctly accept or reject interaction.  Upon every significant user interaction, they must run a check against their stored state to approve or deny the interaction.  This check must be run not only against JS event handlers, but also any default actions that are "dangerous" for their content (e.g. form submission, link navigation) -- probably introducing new active event listeners and preventDefault() calls.  They must be cautious to ensure all new/changed functionality in the iframed content correctly verifies against the stored state.

I think there are other non-IO approaches we could consider that would require less work from the developer, reduce the potential for mistakes, and be more ergonomic.  Brainstorming a couple rough starts:

1. Provide a new property on Event that indicates whether the target passes the visibility test.  This would eliminate the need to maintain state.
2. Similarly, could instead have the property reflect only whether the hosting iframe passes the visibility test rather than the specific element target.
3. Provide a means for the iframed document to request event firing and default actions only when the visibility test passes.  This would both eliminate the need to maintain state, as well as eliminate the need to check trustworthiness in script.
4. Provide an option on event registration to only call that specific handler when the visibility test passes.  e.g. payBtn.addEventListener("click", payMoney, {requireVisibilityInMilliseconds: 200})
5. Provide a document.isVisible() or document.isVisible(element) API, which can be called on demand to check visibility.
6. Build this as a separate API, VisibilityObserver.  Many of the scenarios cited ("like" button, payment widget) probably have no interest in intersection testing but do want to protect against click-jacking.  Receiving events as the user scrolls the "like" buttons in and out of view is unproductive for this purpose - I only want to know when the isVisible flag changes.

As a side benefit in 1-5, if no "delay" is required the visibility test could optionally be performed only on-demand when interaction occurs rather than constantly monitoring state.  If delay is required then the UA would need to monitor state, but at least the content author would be relieved of that task.

If I've missed a key point that makes intersection a key factor in this problem, I think it would be good to add a note in the explainer on "Why IntersectionObserver?" to explain the relationship between click-jacking and intersection.

On the other hand, I can imagine how a visibility flag on IO might be useful for ad tracking ("Has the ad been in the viewport AND fully visible for at least 500ms?").  If this is actually a key scenario that we are seeking to solve simultaneously then it should be detailed in the explainer, so we are validating proposed solutions against the needs of the full set of target scenarios.

**Comment by szager-chromium** [comment]
I disagree that modifying the event handler registration interface is more ergonomic than maintaining a "visible since" state variable. It's a much more invasive spec change, and it requires modifying every event registration in user code.

Anyways, IntersectionObserver is already used in ~20% of all page loads in Chrome:

https://www.chromestatus.com/metrics/feature/timeline/popularity/1368

The API is very well know to developers, and V2 is a pretty small iterative change, so I don't really think ergonomics is much of an issue.

I strongly dislike #5, since it implies/encourages polling.

The reason why I think IntersectionObserver is a good fit is that it only makes sense to measure visibility when the target element is scrolled into the viewport. If VisibilityObserver were implemented as a separate class (#6), then the implementation would presumably first do an intersection test to determine whether it's actually necessary to determine visibility. That's redundant with what IntersectionObserver already does.

The code in the demo/ part of the repository demonstrates, I think, how easy it is to use the current proposal in the context of an iframe handling an input event:

https://github.com/szager-chromium/IntersectionObserver/blob/89e207c8b3ef82d26aac2cf6b313a8cf0399e87f/demo/cashbomb/cashbomb.html#L109
